gcp/workers/alias/upstream_hierarchy_computation.py

```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env python3
+# Copyright 2025 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""OSV Upstream relation computation."""
+
+import datetime
+from google.cloud import ndb
+
+import osv
+import osv.logs
+import json
+
+
+def compute_upstream_hierarchy(target_bug: osv.UpstreamGroup,
+                               bug_groups) -> dict[str, set[str]]:
+  """Computes all upstream vulnerabilities for the given bug ID.
+  The returned list contains all of the bug IDs that are upstream of the
+  target bug ID, including transitive upstreams in a map hierarchy.
+  bug_group:
+        { db_id: bug id
+          upstream_ids: str[bug_ids]
+          last_modified_date}
+  """
+  visited = set()
+  upstream_map = {}
+  to_visit = set([target_bug.db_id])
+  while to_visit:
+    bug_id = to_visit.pop()
+    if bug_id in visited:
+      continue
+    visited.add(bug_id)
+    bug = bug_groups.filter(osv.UpstreamGroup.db_id == bug_id).get()
+    if bug is None:
+      continue
+
+    upstreams = set(bug.upstream_ids)
+    if not upstreams:
+      continue
+    for upstream in upstreams:
+      if upstream not in visited and upstream not in to_visit:
+        to_visit.add(upstream)
+      else:
+        if bug_id not in upstream_map:
+          upstream_map[bug_id] = set([upstream])
+        else:
+          upstream_map[bug_id].add(upstream)
+      upstream_map[bug_id] = upstreams
+      to_visit.update(upstreams - visited)
+  for k, v in upstream_map.items():
+    if k is target_bug.db_id:
+      continue
+    upstream_map[target_bug.db_id] = upstream_map[target_bug.db_id] - v
+  if target_bug.upstream_hierarchy == upstream_map:
+    return upstream_map
+  if upstream_map:
+    target_bug.upstream_hierarchy = json.dumps(
+        upstream_map, default=set_default)
+    target_bug.last_modified = datetime.datetime.now()
+    target_bug.put()
+  return upstream_map
```

Comment:
do we still need this return statement? it seems we are not using it from the caller